### PR TITLE
[AP-3445] Providers can submit more accurate income data

### PR DESCRIPTION
**Pull request by cpjmcquillan**
Before, when providers with employment and bank upload permissions submitted
income data for their client, they could only specify which transactions their
client received.

This data is not granular enough, and makes it impossible to obtain a financial
eligibility assessment (CFE) result for the client.

As part of ongoing work to iterate on the bank upload journey, providers are now
able to tell us how much (amount), and how often (frequency) their client
receives these income payments.

Next steps will include:

- ensuring providers can select the cash transactions
- ensuring cash transactions are updated when regular payments are updated
- replicating this functionality for the outgoings page
- updating the frequencies to reflect the designs
- allowing providers to save and come back later

This user journey is restricted by feature flag at the flow and controller level.

## Video

https://user-images.githubusercontent.com/25187547/189706162-a35dbbde-3424-472c-af55-1514e2c89ae7.mp4

[Link to story](https://dsdmoj.atlassian.net/browse/AP-3445)

### CODE REVIEW

File: app/models/regular_payment.rb
Comment (cpjmcquillan) [inline app/models/regular_payment.rb:2]: These frequencies need updating once they are confirmed by design.

File: db/schema.rb
Comment (cpjmcquillan) [inline db/schema.rb:693]: It looks like these fields are nullable, but I didn't remove the constraint.

Has anyone run in to this before?

File: spec/forms/providers/means/identify_income_form_spec.rb
Comment (cpjmcquillan) [inline spec/forms/providers/means/identify_income_form_spec.rb:39]: I constructed params like this to replicate how they might be submitted from the form in a real browser.

File: app/models/regular_payment.rb
Comment (cpjmcquillan) [reply app/models/regular_payment.rb:2]: Covered by https://dsdmoj.atlassian.net/browse/AP-3459

File: db/schema.rb
Comment (cpjmcquillan) [reply db/schema.rb:23]: I'll revert these changes, but I thought the discussion could be useful.

File: db/migrate/20220902150227_create_regular_payments.rb
Comment (cpjmcquillan) [inline db/migrate/20220902150227_create_regular_payments.rb:3]: `regular_payments` or `regular_transactions`?

In isolation, I think `RegularPayment` makes more sense and models the domain accurately and clearly.

However, given we already have `TransactionType`, maybe `RegularTransaction` would be more helpful in the context of the application?

What do other people think?

Comment (jsugarman) [reply db/migrate/20220902150227_create_regular_payments.rb:3]: We have had a discussion on this already as a team when looking at the first basic ticket and decided to use the regular_transactions `RegularTransaction` naming. That naming has also been used in the [related CFE PR](https://github.com/ministryofjustice/check-financial-eligibility/pull/1076)

File: app/forms/providers/means/identify_income_form.rb
Comment (jsugarman) [inline app/forms/providers/means/identify_income_form.rb:3]: **NOT a blocker** Just from a naming point of view, and following on from the controller renaming comment,  it might be clearer to name it IdentifyRegularIncomesForm. Have followed a similar naming in the CFE branch

**Review by jsugarman — COMMENTED** [review]
looks good, 

have added some comments about naming of the model, controller and form and use of a constant to store the names of the credits, the last being the only real blocker/tbd

File: app/views/providers/means/identify_incomes/show.html.erb
Comment (RoseSAK) [inline app/views/providers/means/identify_incomes/show.html.erb:3]:
```suggestion
    <%= f.govuk_check_boxes_fieldset :income_types, legend: {size: "xl", tag: "h1"} do %>
```

Comment (RoseSAK) [reply app/views/providers/means/identify_incomes/show.html.erb:3]: Would suggest adding this so the legend renders within a h1 tag - otherwise the page is missing a h1

**Review by RoseSAK — COMMENTED** [review]
One final comment, I noticed that there error message that appears when nothing is selected is just the default rails error message: 

Screenshot 2022-09-13 at 16 53 01

This should probably be updated

File: app/forms/providers/means/identify_income_form.rb
Comment (cpjmcquillan) [reply app/forms/providers/means/identify_income_form.rb:3]: In the spirit of the other comment, I'd probably lean toward something like `RegularIncomeForm`.

Does that convey the intent more accurately from your perspective?

File: app/services/flow/flows/provider_capital.rb
Comment (cpjmcquillan) [reply app/services/flow/flows/provider_capital.rb:18]: I agree it isn't very clear at all —when I started the work I just followed the `:identify_types_of_income` that existed already.

I'll leave that flow as is, but I'll refactor the enhanced flow to use more transparent naming.

Probably something like `:regular_incomes`.

How does that sound?

File: config/locales/en/helpers.yml
Comment (cpjmcquillan) [reply config/locales/en/helpers.yml:89]: I added them so we can rely on the form builder's default behaviour to lookup these labels automatically.

File: app/controllers/providers/means/identify_types_of_incomes_controller.rb
Comment (cpjmcquillan) [reply app/controllers/providers/means/identify_types_of_incomes_controller.rb:4]: I'd prefer us to use feature flags as routing constraints, but I didn't have the time to implement it right now.

Given we don't have that (and we don't write unit tests for the flow steps), I think it's appropriate to restrict feature-flagged functionality at the controller level.

File: app/views/providers/means/identify_incomes/show.html.erb
Comment (cpjmcquillan) [reply app/views/providers/means/identify_incomes/show.html.erb:3]: Good spot, thanks!

File: app/forms/providers/means/identify_income_form.rb
Comment (cpjmcquillan) [reply app/forms/providers/means/identify_income_form.rb:6]: Yeah that makes sense, thanks!

File: app/services/flow/flows/provider_capital.rb
Comment (RoseSAK) [inline app/services/flow/flows/provider_capital.rb:44]: Just noticed this - we don't need this conditional as we only need to redirect to the  `:income_summary` page on the Truelayer route - and anyone coming from this page will be on the enhanced bank upload route. So we can just redirect straight to `:means_summaries`

Comment (cpjmcquillan) [reply app/services/flow/flows/provider_capital.rb:44]: Presumably that's the case for the existing `:identify_types_of_income` flow step too?

If it is, let's open a separate PR to fix that!

File: app/forms/providers/means/identify_income_form.rb
Comment (cpjmcquillan) [reply app/forms/providers/means/identify_income_form.rb:6]: I've had a look, but...

I don't think this change is worthwhile, since it necessitates us creating all the transaction type records in all the relevant unit and request specs —and I think the cost of maintaining the list of income types we want to display in the form is low.

If this is something you feel passionate about, we can add a ticket to the backlog and think about it outside the context of this PR.

How does that sound to you?

**Review by jsugarman — CHANGES_REQUESTED** [review]
Need to catchup about this as it is not working because the transaction types themselves are, from what i can tell, not being saved as part of the `RegularIncomeForm` (`IdentifyRegularIncomeForm`, really) so you cannot get back to see them nor add cash transactions. 
Screenshot 2022-09-15 at 16 38 57


I wonder if there should be a join table as well . It may simpler to follow the pattern of the `IdentifyTypesOfIncomesController` more closely

```
has_many :regular_transactions, through: :legal_aid_transaction_type_regular_transactions
```

File: app/services/flow/flows/provider_capital.rb
Comment (RoseSAK) [reply app/services/flow/flows/provider_capital.rb:44]: Yes so since the introduction of this page then the check answers redirect can be updated for that flow step too, since the pages are now separate! 

Separate PR is potentially a good idea to update the `:identify_types_of_income` step as we might need to update/remove tests too 👌

File: config/locales/en/helpers.yml
Comment (RoseSAK) [reply config/locales/en/helpers.yml:89]: That makes sense 👍

File: app/controllers/providers/means/identify_types_of_incomes_controller.rb
Comment (RoseSAK) [reply app/controllers/providers/means/identify_types_of_incomes_controller.rb:4]: The flow step files don't have their own unit test files but we do write rspec request tests which cover redirects for each page based on different feature flags being switched on/off. But happy to go with this method for now if you think it's suitable

File: spec/forms/providers/means/regular_income_form_spec.rb
Comment (jsugarman) [inline spec/forms/providers/means/regular_income_form_spec.rb:7]: I don't feel comfortable with the complete change in style for our rspec tests. 

I think we should keep the existing pattern of using rspec `before/after/around` and `let` to extract out setup as and keep the individual example "thin", as much as possible pending a wider cross-team discussion. 

Personally, i find this style less clear and less open to DRYing out, if appropriate, than existing styles and feels like a backward step to me.

I have already added this [ticket to our trello board](https://trello.com/c/POzgiM2O/127-use-of-before-and-lets-in-specs) to start that off.

File: app/services/flow/flows/provider_capital.rb
Comment (jsugarman) [reply app/services/flow/flows/provider_capital.rb:18]: `identify_regular_incomes` ?!

**Comment by cpjmcquillan** [comment]
@jsugarman I'm not sure any of this necessitates the `Changes requested` badge-of-doom, nor should it be blocking the PR from being merged.

The intention of this PR is to unblock and create a foundation for the rest of the work. From my perspective, it is counter-productive and unhelpful to have ~100 file change PRs that sit around for a long time —since it makes reviewing them so much more difficult. 

The intention of this PR _is not_ to deliver a complete solution that is ready to be released to users (although it can be tested and reviewed by testers/product managers/designers). 

Ideally this PR would have been reviewed and released already.

---

Addressing the specific points you raised:

- > it is not working because the transaction types are not being saved

As mentioned in the commit message:

> Next steps will include:
> - ensuring providers can select the cash transactions

and this is covered in #4279. 

- > I wonder if there should be a join table as well

As you discussed in the [initial ticket](https://dsdmoj.atlassian.net/browse/AP-3426), the preferred solution was to implement a new `regular_transactions` table. I'm not sure what purpose it would serve to duplicate some of that data in an extra table? 

I feel like it would add confusion and make the relationship between a `LegalAidApplication` and its `regular_transactions` and `transaction_type` unnecessarily complicated. 

We would need to maintain a single data point across 2 tables (`regular_transactions` and `legal_aid_application_transaction_types`). I'm not sure what problem that would solve in the long term.

Currently, you could define a relationship between a `LegalAidApplication` and its `transaction_types` like this:

```ruby
class LegalAidApplication < ApplicationRecord
  has_many :regular_transactions
  has_many :transaction_types, through: :regular_transactions
end
```

Ultimately, as we migrate from the existing flow to the enhanced bank upload flow, we will likely define that relationship. We can't currently though, because that relationship is already defined through the `legal_aid_application_transaction_types` table. 

- > It may simpler to follow the pattern of the IdentifyTypesOfIncomesController more closely

I don't agree with this. In my opinion that controller has far too many responsibilities and I think we should be aiming to improve the state of the codebase as we go, not contribute to further controller bloat. It might be "simpler" in the sense we can copy-paste code to get a working solution, but at what maintenance cost?

---

I hope that clarified things for you, and I hope we're able to move this PR forwards so that we can unblock the next steps and continue to iterate!

If you have any other questions, let me know and perhaps we can pair to discuss them?

**Comment by jsugarman** [comment]
Yes, let's discuss in catchup I booked.

> I'm not sure any of this necessitates the Changes requested badge-of-doom..
I personally do not see it as a badge of doom but an indicator that something needs to be discussed addressed prior to merging. For me those things are primiarily

- I think we should be saving transaction types as part of this page/form's functionality so we have that relationship between application and regular "income" transaction in place. 
- the change to rspec test styling that we have discussed before

File: app/forms/providers/means/regular_income_form.rb
Comment (jsugarman) [inline app/forms/providers/means/regular_income_form.rb:63]: add a join table record for that transaction_type if it does not exist already at this point perhaps?!

**Comment by cpjmcquillan** [comment]
Closing this in favour of a new PR because the scope of the branch has changed significantly and much of the conversation here is no longer relevant.

**Pull request closed without merging**
